src/generated/resources/openapi.yaml

```diff
@@ -733,6 +763,30 @@ components:
           maxLength: 256
           minLength: 1
           type: string
+        credentials:
+          description: "The credentials for the Kafka cluster, or null if no authentication\
+            \ is required"
+          type: object
+          allOf:
+          - $ref: "#/components/schemas/Credentials"
```

Comment:
TLDR: Including `allOf` here generates the API client code with `credentials` as a generic `object` rather than the more strongly typed union object. Let's keep it this way while the incessant bugs in the typescript-fetch template are fixed, at the cost of a bit of developer unfriendliness in trying to determine whether the opaque `object` is a `BasicCredentials` or `ApiKeyAndSecret`.

Upon running `npx gulp apigen format` in the confluentinc/vscode repo after copying over the OpenAPI spec from this branch, I see the following diff for `src/clients/sidecar/models/KafkaClusterConfig.ts` (out of a few other changed files too). You'll see that the type for `credentials` in the generated TypeScript interface is `object | null`. 
<details>
<summary> Git diff for `src/clients/sidecar/models/KafkaClusterConfig.ts` </summary>

```


diff --git a/src/clients/sidecar/models/KafkaClusterConfig.ts b/src/clients/sidecar/models/KafkaClusterConfig.ts
index d9fb41b..d1f0695 100644
--- a/src/clients/sidecar/models/KafkaClusterConfig.ts
+++ b/src/clients/sidecar/models/KafkaClusterConfig.ts
@@ -31,6 +31,24 @@ export interface KafkaClusterConfig {
    * @memberof KafkaClusterConfig
    */
   bootstrap_servers: string;
+  /**
+   * The credentials for the Kafka cluster, or null if no authentication is required
+   * @type {object}
+   * @memberof KafkaClusterConfig
+   */
+  credentials?: object | null;
+  /**
+   * Whether to communicate with the Kafka cluster over TLS/SSL. Defaults to 'true', but set to 'false' when the Kafka cluster does not support TLS/SSL.
+   * @type {boolean}
+   * @memberof KafkaClusterConfig
+   */
+  ssl?: boolean | null;
+  /**
+   * Whether to verify the Kafka cluster certificates. Defaults to 'true', but set to 'false' when the Kafka cluster has self-signed certificates.
+   * @type {boolean}
+   * @memberof KafkaClusterConfig
+   */
+  verify_ssl_certificates?: boolean | null;
 }
 
 /**
@@ -55,6 +73,10 @@ export function KafkaClusterConfigFromJSONTyped(
   return {
     id: json["id"] == null ? undefined : json["id"],
     bootstrap_servers: json["bootstrap_servers"],
+    credentials: json["credentials"] == null ? undefined : json["credentials"],
+    ssl: json["ssl"] == null ? undefined : json["ssl"],
+    verify_ssl_certificates:
+      json["verify_ssl_certificates"] == null ? undefined : json["verify_ssl_certificates"],
   };
 }
 
@@ -65,5 +87,8 @@ export function KafkaClusterConfigToJSON(value?: KafkaClusterConfig | null): any
   return {
     id: value["id"],
     bootstrap_servers: value["bootstrap_servers"],
+    credentials: value["credentials"],
+    ssl: value["ssl"],
+    verify_ssl_certificates: value["verify_ssl_certificates"],
   };
 }

```

</details>

To actually declare the type of `credentials` as a union of `BasicCredentials` and `ApiKeyAndSecret`, we'll need to remove the `allOf` from the spec above (using some bespoke parameter in the `@Schema` annotation somewhere). 

So, I did exactly that, removed the `allOf` field from both `KafkaClusterConfig` and `SchemaRegistryClusterConfig`, then generated the clients again using `npx gulp apigen format`. Well, this opened a can of worms. Read on..
 
confluentinc/vscode uses the [typescript-fetch](https://github.com/confluentinc/vscode/blob/main/Gulpfile.js#L672) template to generate the API client code, and we've currently [pinned its version to 7.7.0](https://github.com/confluentinc/vscode/blob/main/Gulpfile.js#L630). The generated type for `credentials` was how I'd expected it to be (see below), but unfortunately, there was a bug in one of the functions in the generated code: "Function lacks ending return statement and return type does not include 'undefined'". 

```typescript
// In src/clients/sidecar/models/KafkaClusterConfig.ts
...
  /**
   *
   * @type {KafkaClusterConfigCredentials}
   * @memberof KafkaClusterConfig
   */
  credentials?: KafkaClusterConfigCredentials | null;
...
```

```
// In src/clients/sidecar/models/KafkaClusterConfigCredentials.ts
...
/**
 * @type KafkaClusterConfigCredentials
 * The credentials for the Kafka cluster, or null if no authentication is required
 * @export
 */
export type KafkaClusterConfigCredentials = ApiKeyAndSecret | BasicCredentials;
...
```

So, I upgraded the openapi-generator-version to the latest stable version 7.9.0 in the hopes that this bug will disappear and regenerated the API clients (370 file diff) with the sans-`allOf` OpenAPI spec. Unfortunately, there was yet another bug in the generated code. It was [fixed](https://github.com/OpenAPITools/openapi-generator/pull/20046) just last week and will likely be a while before it makes it into a stable release.

All this to say, let's keep this spec as is and wait to see if the next release of openapi-generator poses issues.